Author: Rene Kita <mail@rkta.de>
Date: Fri Aug 11 20:04:51 CEST 2023
Subject: replace deprecated bcopy

WIP: Still needs changes in configure.ac, but can't run autoconf.
---
 Str.c       |   12 ++++++------
 buffer.c    |    2 +-
 config.h.in |    1 -
 cookie.c    |    2 +-
 file.c      |   34 +++++++++++++++++-----------------
 fm.h        |    4 ----
 form.c      |   12 ++++++------
 frame.c     |    8 ++++----
 func.c      |    8 +++-----
 indep.c     |   27 ---------------------------
 istream.c   |    4 ++--
 main.c      |    4 ++--
 matrix.h    |    2 +-
 table.c     |    2 +-
 url.c       |    4 ++--
 15 files changed, 46 insertions(+), 80 deletions(-)

--- a/Str.c
+++ b/Str.c
@@ -89,7 +89,7 @@ Strnew_charp(const char *p)
 	exit(1);
     x->area_size = n;
     x->length = len;
-    bcopy(p, x->ptr, len);
+    memmove(x->ptr, p, len);
     x->ptr[x->length] = '\0';
     return x;
 }
@@ -130,7 +130,7 @@ Strnew_charp_n(const char *p, int n)
 	exit(1);
     x->area_size = n + 1;
     x->length = len;
-    bcopy(p, x->ptr, len);
+    memmove(x->ptr, p, len);
     x->ptr[x->length] = '\0';
     return x;
 }
@@ -169,7 +169,7 @@ Strcopy(Str x, Str y)
 	    exit(1);
 	x->area_size = y->length + 1;
     }
-    bcopy(y->ptr, x->ptr, y->length + 1);
+    memmove(x->ptr, y->ptr, y->length + 1);
     x->length = y->length;
 }
 
@@ -193,7 +193,7 @@ Strcopy_charp(Str x, const char *y)
 	    exit(1);
 	x->area_size = len + 1;
     }
-    bcopy(y, x->ptr, len);
+    memmove(x->ptr, y, len);
     x->ptr[len] = '\0';
     x->length = len;
 }
@@ -217,7 +217,7 @@ Strcopy_charp_n(Str x, const char *y, in
 	    exit(1);
 	x->area_size = len + 1;
     }
-    bcopy(y, x->ptr, len);
+    memmove(x->ptr, y, len);
     x->ptr[len] = '\0';
     x->length = len;
 }
@@ -248,7 +248,7 @@ Strcat_charp_n(Str x, const char *y, int
 	    exit(1);
 	x->area_size = newlen;
     }
-    bcopy(y, &x->ptr[x->length], n);
+    memmove(&x->ptr[x->length], y, n);
     x->length += n;
     x->ptr[x->length] = '\0';
 }
--- a/buffer.c
+++ b/buffer.c
@@ -591,7 +591,7 @@ void
 copyBuffer(Buffer *a, Buffer *b)
 {
     readBufferCache(b);
-    bcopy(b, a, sizeof(Buffer));
+    memmove(a, b, sizeof(Buffer));
 }
 
 Buffer *
--- a/cookie.c
+++ b/cookie.c
@@ -222,7 +222,7 @@ find_cookie(ParsedURL *pu)
 	    if (p1)
 		continue;
 	    p1 = New(struct cookie);
-	    bcopy(p, p1, sizeof(struct cookie));
+	    memmove(p1, p, sizeof(struct cookie));
 	    p1->next = fco;
 	    fco = p1;
 	    if (p1->version > version)
--- a/file.c
+++ b/file.c
@@ -2640,8 +2640,8 @@ set_breakpoint(struct readbuffer *obuf,
     if (!obuf->bp.init_flag)
 	return;
 
-    bcopy((void *)&obuf->anchor, (void *)&obuf->bp.anchor,
-	  sizeof(obuf->anchor));
+    memmove((void *)&obuf->bp.anchor, (void *)&obuf->anchor,
+	    sizeof(obuf->anchor));
     obuf->bp.img_alt = obuf->img_alt;
     obuf->bp.input_alt = obuf->input_alt;
     obuf->bp.in_bold = obuf->in_bold;
@@ -2658,8 +2658,8 @@ static void
 back_to_breakpoint(struct readbuffer *obuf)
 {
     obuf->flag = obuf->bp.flag;
-    bcopy((void *)&obuf->bp.anchor, (void *)&obuf->anchor,
-	  sizeof(obuf->anchor));
+    memmove((void *)&obuf->anchor, (void *)&obuf->bp.anchor,
+	    sizeof(obuf->anchor));
     obuf->img_alt = obuf->bp.img_alt;
     obuf->input_alt = obuf->bp.input_alt;
     obuf->in_bold = obuf->bp.in_bold;
@@ -2964,7 +2964,7 @@ flushline(struct html_feed_environ *h_en
 
 	if (obuf->bp.pos == obuf->pos && tp <= ep &&
 	    tp > line->ptr && tp[-1] == ' ') {
-	    bcopy(tp, tp - 1, ep - tp + 1);
+	    memmove(tp - 1, tp, ep - tp + 1);
 	    line->length--;
 	    obuf->pos--;
 	}
@@ -3213,8 +3213,8 @@ close_effect0(struct readbuffer *obuf, i
     }
     if (i >= 0) {
 	obuf->tag_sp--;
-	bcopy(&obuf->tag_stack[i + 1], &obuf->tag_stack[i],
-	      (obuf->tag_sp - i) * sizeof(struct cmdtable *));
+	memmove(&obuf->tag_stack[i], &obuf->tag_stack[i + 1],
+		(obuf->tag_sp - i) * sizeof(struct cmdtable *));
 	return 1;
     }
     else if ((p = has_hidden_link(obuf, cmd)) != NULL) {
@@ -3250,8 +3250,8 @@ close_anchor(struct html_feed_environ *h
 	    else {
 		if (i >= 0) {
 		    obuf->tag_sp--;
-		    bcopy(&obuf->tag_stack[i + 1], &obuf->tag_stack[i],
-			  (obuf->tag_sp - i) * sizeof(struct cmdtable *));
+		    memmove(&obuf->tag_stack[i], &obuf->tag_stack[i + 1],
+			    (obuf->tag_sp - i) * sizeof(struct cmdtable *));
 		}
 		else {
 		    passthrough(obuf, p, 1);
@@ -3275,8 +3275,8 @@ void
 save_fonteffect(struct html_feed_environ *h_env, struct readbuffer *obuf)
 {
     if (obuf->fontstat_sp < FONT_STACK_SIZE)
-	bcopy(obuf->fontstat, obuf->fontstat_stack[obuf->fontstat_sp],
-	      FONTSTAT_SIZE);
+	memmove(obuf->fontstat_stack[obuf->fontstat_sp], obuf->fontstat,
+		FONTSTAT_SIZE);
     if (obuf->fontstat_sp < INT_MAX)
 	obuf->fontstat_sp++;
     if (obuf->in_bold)
@@ -3298,8 +3298,8 @@ restore_fonteffect(struct html_feed_envi
     if (obuf->fontstat_sp > 0)
 	obuf->fontstat_sp--;
     if (obuf->fontstat_sp < FONT_STACK_SIZE)
-	bcopy(obuf->fontstat_stack[obuf->fontstat_sp], obuf->fontstat,
-	      FONTSTAT_SIZE);
+	memmove(obuf->fontstat, obuf->fontstat_stack[obuf->fontstat_sp],
+		FONTSTAT_SIZE);
     if (obuf->in_bold)
 	push_tag(obuf, "<b>", HTML_B);
     if (obuf->in_italic)
@@ -6913,7 +6913,7 @@ addnewline(Buffer *buf, char *line, Line
     if (pos > 0) {
 	s = allocStr(line, pos);
 	p = NewAtom_N(Lineprop, pos);
-	bcopy(prop, p, pos * sizeof(Lineprop));
+	memmove(p, prop, pos * sizeof(Lineprop));
     }
     else {
 	s = NullLine;
@@ -6922,7 +6922,7 @@ addnewline(Buffer *buf, char *line, Line
 #ifdef USE_ANSI_COLOR
     if (pos > 0 && color) {
 	c = NewAtom_N(Linecolor, pos);
-	bcopy(color, c, pos * sizeof(Linecolor));
+	memmove(c, color, pos * sizeof(Linecolor));
     }
     else {
 	c = NULL;
@@ -8338,7 +8338,7 @@ save2tmp(URLFile uf, const char *tmpf)
 	/* fclose(f); */
 	return -1;
     }
-    bcopy(AbortLoading, env_bak, sizeof(JMP_BUF));
+    memmove(env_bak, AbortLoading, sizeof(JMP_BUF));
     if (SETJMP(AbortLoading) != 0) {
 	goto _end;
     }
@@ -8383,7 +8383,7 @@ save2tmp(URLFile uf, const char *tmpf)
 	}
     }
   _end:
-    bcopy(env_bak, AbortLoading, sizeof(JMP_BUF));
+    memmove(AbortLoading, env_bak, sizeof(JMP_BUF));
     TRAP_OFF;
     xfree(buf);
     fclose(ff);
--- a/form.c
+++ b/form.c
@@ -324,8 +324,8 @@ form_update_line(Line *line, char **str,
     buf = New_N(char, len + 1);
     buf[len] = '\0';
     prop = New_N(Lineprop, len);
-    bcopy(line->lineBuf, buf, spos * sizeof(char));
-    bcopy(line->propBuf, prop, spos * sizeof(Lineprop));
+    memmove(buf, line->lineBuf, spos * sizeof(char));
+    memmove(prop, line->propBuf, spos * sizeof(Lineprop));
 
     effect = CharEffect(line->propBuf[spos]);
     for (p = *str, w = 0, pos = spos; *p && w < width;) {
@@ -402,10 +402,10 @@ form_update_line(Line *line, char **str,
     }
     *str = p;
 
-    bcopy((void *)&line->lineBuf[epos], (void *)&buf[pos],
-	  (line->len - epos) * sizeof(char));
-    bcopy((void *)&line->propBuf[epos], (void *)&prop[pos],
-	  (line->len - epos) * sizeof(Lineprop));
+    memmove((void *)&buf[pos], (void *)&line->lineBuf[epos],
+	    (line->len - epos) * sizeof(char));
+    memmove((void *)&prop[pos], (void *)&line->propBuf[epos],
+	    (line->len - epos) * sizeof(Lineprop));
     line->lineBuf = buf;
     line->propBuf = prop;
     line->len = len;
--- a/frame.c
+++ b/frame.c
@@ -169,7 +169,7 @@ copyFrame(struct frame_body *ob)
     struct frame_body *rb;
 
     rb = New(struct frame_body);
-    bcopy(ob, rb, sizeof(struct frame_body));
+    memmove(rb, ob, sizeof(struct frame_body));
     return rb;
 }
 
@@ -181,11 +181,11 @@ copyFrameSet(struct frameset *of)
 
     rf = New(struct frameset);
     n = of->col * of->row;
-    bcopy(of, rf, sizeof(struct frameset));
+    memmove(rf, of, sizeof(struct frameset));
     rf->width = New_N(char *, rf->col);
-    bcopy(of->width, rf->width, sizeof(char *) * rf->col);
+    memmove(rf->width, of->width, sizeof(char *) * rf->col);
     rf->height = New_N(char *, rf->row);
-    bcopy(of->height, rf->height, sizeof(char *) * rf->row);
+    memmove(rf->height, of->height, sizeof(char *) * rf->row);
     rf->frame = New_N(union frameset_element, n);
     while (n) {
 	n--;
--- a/func.c
+++ b/func.c
@@ -685,12 +685,10 @@ initMouseAction(void)
 {
     FILE *mf;
 
-    bcopy(&default_mouse_action, &mouse_action,
-	  sizeof(default_mouse_action));
+    memmove(&mouse_action, &default_mouse_action, sizeof(default_mouse_action));
     mouse_action.lastline_map[0] = New_N(MouseActionMap, 6);
-    bcopy(&default_lastline_action,
-	  mouse_action.lastline_map[0],
-	  sizeof(default_lastline_action));
+    memmove(mouse_action.lastline_map[0], &default_lastline_action,
+	    sizeof(default_lastline_action));
     {
 #ifdef USE_M17N
 	int w = 0;
--- a/istream.c
+++ b/istream.c
@@ -57,7 +57,7 @@ buffer_read(StreamBuffer sb, unsigned ch
     if (len > 0) {
 	if (len > count)
 	    len = count;
-	bcopy((const void *)&sb->buf[sb->cur], obuf, len);
+	memmove(obuf, (const void *)&sb->buf[sb->cur], len);
 	sb->cur += len;
     }
     return len;
@@ -440,7 +440,7 @@ ssl_check_cert_ident(X509 * x, char *hos
 		    char *asn = GC_MALLOC(sl + 1);
 		    if (!asn)
 			exit(1);
-		    bcopy(sn, asn, sl);
+		    memmove(asn, sn, sl);
 		    asn[sl] = '\0';
 
 		    if (!seen_dnsname)
--- a/main.c
+++ b/main.c
@@ -4776,7 +4776,7 @@ _peekURL(int only_img)
 #ifdef USE_M17N
     s = checkType(s, &pp, NULL);
     p = NewAtom_N(Lineprop, s->length);
-    bcopy(pp, p, s->length * sizeof(Lineprop));
+    memmove(p, pp, s->length * sizeof(Lineprop));
 #endif
   disp:
     n = searchKeyNum();
@@ -4835,7 +4835,7 @@ DEFUN(curURL, PEEK, "Show current addres
 #ifdef USE_M17N
 	s = checkType(s, &pp, NULL);
 	p = NewAtom_N(Lineprop, s->length);
-	bcopy(pp, p, s->length * sizeof(Lineprop));
+	memmove(p, pp, s->length * sizeof(Lineprop));
 #endif
     }
     n = searchKeyNum();
--- a/table.c
+++ b/table.c
@@ -3386,7 +3386,7 @@ pushTable(struct table *tbl, struct tabl
 	    tbl->tables_size = MAX_TABLE_N_LIMIT;
 	tmp = New_N(struct table_in, tbl->tables_size);
 	if (tbl->tables)
-	    bcopy(tbl->tables, tmp, tbl->ntable * sizeof(struct table_in));
+	    memmove(tmp, tbl->tables, tbl->ntable * sizeof(struct table_in));
 	tbl->tables = tmp;
     }
 
--- a/url.c
+++ b/url.c
@@ -645,7 +645,7 @@ openSocket(char *const hostname,
     if (regexMatch(hostname, -1, 1)) {
 	sscanf(hostname, "%d.%d.%d.%d", &a1, &a2, &a3, &a4);
 	adr = htonl((a1 << 24) | (a2 << 16) | (a3 << 8) | a4);
-	bcopy((void *)&adr, (void *)&hostaddr.sin_addr, sizeof(long));
+	memmove((void *)&hostaddr.sin_addr, (void *)&adr, sizeof(long));
 	hostaddr.sin_family = AF_INET;
 	hostaddr.sin_port = s_port;
 	if (fmInitialized) {
@@ -679,7 +679,7 @@ openSocket(char *const hostname,
 	hostaddr.sin_family = AF_INET;
 	hostaddr.sin_port = s_port;
 	for (h_addr_list = entry->h_addr_list; *h_addr_list; h_addr_list++) {
-	    bcopy(h_addr_list[0], &hostaddr.sin_addr, entry->h_length);
+	    memmove(&hostaddr.sin_addr, h_addr_list[0], entry->h_length);
 #ifdef SOCK_DEBUG
 	    adr = ntohl(*(long *)&hostaddr.sin_addr);
 	    sock_log("openSocket: connecting %d.%d.%d.%d\n",
--- a/fm.h
+++ b/fm.h
@@ -77,10 +77,6 @@
 #include "terms.h"
 #include "istream.h"
 
-#ifndef HAVE_BCOPY
-void bcopy(const void *, void *, int);
-void bzero(void *, int);
-#endif				/* HAVE_BCOPY */
 #ifdef __EMX__
 #include <strings.h>		/* for bzero() and bcopy() */
 #endif
--- a/indep.c
+++ b/indep.c
@@ -73,33 +73,6 @@ strtoclen(const char *s)
 #endif
 }
 
-#ifndef HAVE_BCOPY
-void
-bcopy(const void *src, void *dest, int len)
-{
-    int i;
-    if (src == dest)
-	return;
-    if (src < dest) {
-	for (i = len - 1; i >= 0; i--)
-	    ((char *)dest)[i] = ((const char *)src)[i];
-    }
-    else {			/* src > dest */
-	for (i = 0; i < len; i++)
-	    ((char *)dest)[i] = ((const char *)src)[i];
-    }
-}
-
-void
-bzero(void *ptr, int len)
-{
-    int i;
-    char *p = ptr;
-    for (i = 0; i < len; i++)
-	*(p++) = 0;
-}
-#endif				/* not HAVE_BCOPY */
-
 char *
 allocStr(const char *s, int len)
 {
--- a/matrix.h
+++ b/matrix.h
@@ -41,7 +41,7 @@ typedef struct vector *Vector;
 
 #define m_entry(m,i,j) (M_VAL(m,i,j))
 #define v_entry(v,i) (V_VAL(v,i))
-#define m_copy(m1,m2) (bcopy((m1)->me,(m2)->me,(m1)->dim*(m1)->dim*sizeof(double)))
+#define m_copy(m1,m2) (memmove((m2)->me,(m1)->me,(m1)->dim*(m1)->dim*sizeof(double)))
 #define v_free(v) ((v)=NULL)
 #define m_free(m) ((m)=NULL)
 #define px_free(px) ((px)=NULL)
--- a/config.h.in
+++ b/config.h.in
@@ -129,7 +129,6 @@ typedef long clen_t;
 #undef HAVE_STRCASESTR
 #undef HAVE_STRCHR
 #undef HAVE_STRERROR
-#undef HAVE_BCOPY
 #undef HAVE_WAITPID
 #undef HAVE_WAIT3
 #undef HAVE_STRFTIME
